### PR TITLE
feat: new binarypbindexer based on memmap

### PR DESCRIPTION
**Pull request by hanxiao**
related to #929

### CODE REVIEW

File: jina/executors/indexers/keyvalue.py [hunk 1]
```diff
@@ -1,161 +1,72 @@
...
+    def query(self, key: int) -> Optional['jina_pb2.Document']:
 
-        :param obj: an object
-        """
-        for k in keys:
-            self.write_handler.write(k.SerializeToString() + __binary_delimiter__)
-            self._size += 1
+        pos_info = self.query_handler.get(key, None)
+        if pos_info is not None:
+            p, r, l = pos_info
```
Comment (JoanFM) [inline jina/executors/indexers/keyvalue.py:64]: Could it be possible to have more explicative variable names?

File: jina/executors/indexers/__init__.py [hunk 2]
```diff
@@ -2,14 +2,17 @@
...
+from typing import Tuple, Union, List, Iterator, Optional
 
 import numpy as np
 
 from .. import BaseExecutor
 from ..compound import CompoundExecutor
 from ...helper import call_obj_fn, cached_property, get_readable_size
 
+if False:
+    from ...proto import jina_pb2
```
Comment (maximilianwerk) [inline jina/executors/indexers/__init__.py:14]: This is added for type hinting, right?

File: jina/executors/indexers/__init__.py [hunk 3]
```diff
@@ -2,14 +2,17 @@
...
+
 
 class BaseIndexer(BaseExecutor):
     """``BaseIndexer`` stores and searches with vectors.
```
Comment (maximilianwerk) [inline jina/executors/indexers/__init__.py:18]: When you explicitly take out the vector docstring below, I'd also rewrite this to something like

"base class for storing and searching any kind of data strcuture"

File: jina/executors/indexers/__init__.py [hunk 4]
```diff
@@ -190,6 +199,17 @@ class BaseKVIndexer(BaseIndexer):
     It can be used to tell whether an indexer is key-value indexer, via ``isinstance(a, BaseKVIndexer)``
     """
 
+    def add(self, docs: Iterator['jina_pb2.Document'], *args, **kwargs):
```
Comment (maximilianwerk) [inline jina/executors/indexers/__init__.py:202]: I'd add a docstring, which hints, that the key will always be the `id` of the documents.

File: jina/executors/indexers/keyvalue.py [hunk 5]
```diff
@@ -1,161 +1,72 @@
...
 
 from . import BaseKVIndexer
-from ... import __binary_delimiter__
-from ...helper import call_obj_fn
 from ...proto import jina_pb2
 
 
 class BasePbIndexer(BaseKVIndexer):
-    """Storing and querying protobuf chunk/document using gzip and Python dict. """
+    """Depreciated, just for back-compat and for keeping the inheritance order unchanged"""
```
Comment (maximilianwerk) [inline jina/executors/indexers/keyvalue.py:14]: When would we remove this, if not now?

File: jina/executors/indexers/keyvalue.py [hunk 6]
```diff
@@ -1,161 +1,72 @@
...
-    def _add(self, keys: Iterator['jina_pb2.Document'], *args, **kwargs):
-        """Add a object to the indexer
+    def query(self, key: int) -> Optional['jina_pb2.Document']:
 
-        :param obj: an object
-        """
-        for k in keys:
-            self.write_handler.write(k.SerializeToString() + __binary_delimiter__)
-            self._size += 1
+        pos_info = self.query_handler.get(key, None)
```
Comment (maximilianwerk) [inline jina/executors/indexers/keyvalue.py:62]: Personally, I don't like the magic, how this calls `get_query_handler`. I have to look at how and why this is implemented as it is. From my point of view it should either be a `@property` (if the invocation is cheap) or a direct call to the `get_query_handler` if invocation is expensive.

File: tests/unit/executors/indexers/test_shelfindexer.py [hunk 7]
```diff
@@ -1,22 +1,23 @@
-from jina.executors.indexers.keyvalue import ShelfPbIndexer
+from jina.executors.indexers.keyvalue import BinaryPbIndexer
```
Comment (maximilianwerk) [inline tests/unit/executors/indexers/test_shelfindexer.py:1]: Filename should be renamed to `test_binary_indexer.py`

File: tests/unit/executors/indexers/test_shelfindexer.py [hunk 8]
```diff
@@ -31,3 +32,5 @@ def validate(req):
 
     with f:
         f.search([d], output_fn=validate)
+
+test_binarypb_in_flow()
```
Comment (maximilianwerk) [inline tests/unit/executors/indexers/test_shelfindexer.py:36]: This invocation is not necessary if you run it via `pytest`. If you just want to test this test locally, you can also do `pytest tests/unit/executors/indexers/test_shelfindexer.py::test_binarypb_in_flow` which only invokes exactly this test.

File: jina/executors/indexers/__init__.py (hunk 2 above)
Comment (JoanFM) [reply jina/executors/indexers/__init__.py:14]: yes!

File: jina/clients/python/request.py [hunk 9]
```diff
@@ -98,7 +99,7 @@ def _generate(data: Union[Iterator['jina_pb2.Document'], Iterator[bytes], Iterat
                 req.queryset.extend([top_k_queryset])
 
         for content in batch:
-            _add_document(request=req, content=content, mode=mode, doc_counter=doc_counter,
+            _add_document(request=req, content=content, mode=mode, doc_counter=doc_counter if override_doc_id else None,
```
Comment (maximilianwerk) [inline jina/clients/python/request.py:102]: That just became impossible to grasp when just skimming the function. Could you put one parameter per line in order to make it more obvious, that something else then just parameter passing is happening here?

File: tests/unit/executors/indexers/test_shelfindexer.py (hunk 8 above)
Comment (hanxiao) [reply tests/unit/executors/indexers/test_shelfindexer.py]: its a draft for easing testing locally

File: jina/executors/indexers/keyvalue.py (hunk 5 above)
Comment (hanxiao) [reply jina/executors/indexers/keyvalue.py:14]: its a draft for testing

**Pull request merged**